pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +264,276 @@ func (p *Pod) Skip() bool {
+		}
+
+		// add the owner ref
+		if err := controllerutil.SetControllerReference(p.Object(), wl, c.Scheme()); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, pod := range podsInGroup.Items {
+		if err := controllerutil.SetOwnerReference(&pod, wl, c.Scheme()); err != nil {
+			return nil, err
+		}
+	}
+
+	return wl, nil
+}
+
+func (p *Pod) FindMatchingWorkloads(ctx context.Context, c client.Client) (*kueue.Workload, []*kueue.Workload, *kueue.WorkloadList, error) {
+	log := ctrl.LoggerFrom(ctx)
+	object := p.Object()
+	groupName := p.GetGroupName()
+
+	if groupName == "" {
+		return jobframework.FindMatchingWorkloads(ctx, c, p, object)
+	}
+
+	// Find a matching workload first if there is one.
+	var toDelete []*kueue.Workload
+	var match *kueue.Workload
+	workloads := &kueue.WorkloadList{}
+
+	if err := c.List(ctx, workloads, client.InNamespace(object.GetNamespace()),
+		client.MatchingFields{WorkloadNameKey: p.GetGroupName()}); err != nil {
+		log.Error(err, "Unable to list related workloads")
+		return nil, nil, nil, err
+	}
+
+	var podsInGroup corev1.PodList
+	if err := c.List(ctx, &podsInGroup, client.MatchingLabels{
+		GroupNameLabel: p.GetGroupName(),
+	}); err != nil {
+		return nil, nil, nil, err
+	}
+
+	jobPodSets, err := p.constructGroupPodSets(podsInGroup)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for i := range workloads.Items {
+		w := &workloads.Items[i]
+		if match == nil && p.equivalentToWorkload(w, jobPodSets) {
+			match = w
+		} else {
+			toDelete = append(toDelete, w)
+		}
+	}
+
+	return match, toDelete, workloads, nil
+}
+
+func (p *Pod) equivalentToWorkload(wl *kueue.Workload, jobPodSets []kueue.PodSet) bool {
+	if wl.GetName() != p.GetGroupName() {
+		return false
+	}
+
+	if len(jobPodSets) != len(wl.Spec.PodSets) {
+		return false
+	}
+
+	for i := range wl.Spec.PodSets {
+		if wl.Spec.PodSets[i].Count != jobPodSets[i].Count {
+			return false
+		}
+		if wl.Spec.PodSets[i].Name != jobPodSets[i].Name {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *Pod) IsComposableJobFinished(ctx context.Context, c client.Client) (metav1.Condition, bool) {
+	log := ctrl.LoggerFrom(ctx)
+	groupName := p.GetGroupName()
+	if groupName == "" {
+		return p.Finished()
+	}
+
+	var podsInGroup corev1.PodList
+	if err := c.List(ctx, &podsInGroup, client.MatchingLabels{
+		GroupNameLabel: groupName,
+	}); err != nil {
+		return metav1.Condition{}, false
+	}
+
+	failedPodCount := 0
+
+	for _, pod := range podsInGroup.Items {
+		if pod.Status.Phase == corev1.PodFailed {
+			failedPodCount++
+		}
+
+		podJob := fromObject(&pod)
+		if _, finished := podJob.Finished(); !finished {
+			return metav1.Condition{}, false
+		}
+	}
+
+	groupTotalCount, err := p.GetGroupTotalCount()
+	if err != nil {
+		log.V(2).Error(err, "Error while checking if pod group is finished", "Pod.Name", p.Name)
+		return metav1.Condition{}, false
+	}
+
+	condition := metav1.Condition{
+		Type:   kueue.WorkloadFinished,
+		Status: metav1.ConditionTrue,
```

Comment:
are we always returning true?
The job should only considered finished if the number of successful pods equals the number of pods for the group.

Reply:
Still pending resolution.

Reply:
Changed in https://github.com/kubernetes-sigs/kueue/pull/1319/commits/fb5a4eb591521e5e3e2d7acfb4b6268a697b52e9